lib/warden/cognito/token_authenticatable_strategy.rb

```diff
@@ -27,13 +27,29 @@ def authenticate!
         fail!(:unknown_user) unless user.present?
         success!(user)
       rescue ::JWT::ExpiredSignature
-        fail!(:token_expired)
+        try_refresh
       rescue StandardError
         fail(:unknown_error)
       end
 
+      def store?
+        false
+      end
+
       private
 
+      def try_refresh
+        username = ::JWT.decode(token, nil, false).first['username']
+        result = CognitoClient.scope(pool_identifier).exchange_token(refresh_token, username).authentication_result
+        fail(:unknown_error) unless result
+
+        @token_decoder = TokenDecoder.new(result.access_token, pool_identifier)
+        cookies['AccessToken'] = result.access_token
+        authenticate!
+      rescue Aws::CognitoIdentityProvider::Errors::ServiceError
+        fail!(:token_expired)
+      end
```

Comment:
Do we know what the possible errors or return values are here? Would be great to have some test coverage. Also for the cookies, does this respect the application settings for encryption / http only?

Reply:
The cookie stuff is kind of weird. The tests use `Rack::MockRequest` which is great because that means the code can be used for all `Rack` apps. However, it makes the actual app code `Rack` specific which is why I had to parse the cookies manually [here](https://github.com/invibelabs/warden-cognito/pull/2/files#diff-63e09bfa87a4806268ff6e0a2e739933263d5ff6bfa3feaeda60194ee08bb350R88). Rails has some convenience in that they parse cookies for you and make a `cookies` object which doesn't exist in Rack. 

All that to say - [`cookies[key]=`](https://github.com/invibelabs/warden-cognito/pull/2/files#diff-63e09bfa87a4806268ff6e0a2e739933263d5ff6bfa3feaeda60194ee08bb350R47) is a Rails thing so I think it should respect all rails things. As long as this code is run in rails and not just rack, `cookies[key]=` should work. (but also makes testing this hard)

RE: Errors - https://docs.aws.amazon.com/sdk-for-ruby/v3/api/Aws/CognitoIdentityProvider/Errors.html is the list of all errors that could happen from the AWS client